# Importing libraries

In [1]:
import numpy as np
import pandas as pd

# 2. Data Quality Issues and data general insights

## Exploring data

I first start by looking at the data using .info() method, so that I have a general understanding of the dataset

In [2]:
homicidios_2018 = pd.read_csv('./dataset/Homicidios_2018.csv', encoding = 'utf-8')
homicidios_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4966 entries, 0 to 4965
Data columns (total 20 columns):
Fecha                 4965 non-null object
Departamento          4965 non-null object
Municipio             4965 non-null object
Día                   4965 non-null object
Hora                  4965 non-null object
Barrio                4965 non-null object
Zona                  4965 non-null object
Clase de sitio        4965 non-null object
Arma empleada         4965 non-null object
Móvil Agresor         4965 non-null object
Móvil Victima         4965 non-null object
Edad                  4965 non-null float64
Sexo                  4965 non-null object
Estado civil          4965 non-null object
País de nacimiento    4965 non-null object
Clase de empleado     4965 non-null object
Profesión             4965 non-null object
Escolaridad           4965 non-null object
Código DANE           4965 non-null float64
Cantidad              4966 non-null int64
dtypes: float64(2), int64(1), ob

Then by using .isna().any(axis=1) or .isnull().any(axis=1) I find there is one row with no information, so that it needs to be deleted from the dataset using .dropna()

In [3]:
homicidios_2018['Municipio'].isna().value_counts()

False    4965
True        1
Name: Municipio, dtype: int64

In [4]:
homicidios_2018[homicidios_2018.isna().any(axis=1) | homicidios_2018.isnull().any(axis=1)]

Fecha Departamento Municipio  Día Hora Barrio Zona Clase de sitio  \
4965   NaN          NaN       NaN  NaN  NaN    NaN  NaN            NaN   

     Arma empleada Móvil Agresor Móvil Victima  Edad Sexo Estado civil  \
4965           NaN           NaN           NaN   NaN  NaN          NaN   

     País de nacimiento Clase de empleado Profesión Escolaridad  Código DANE  \
4965                NaN               NaN       NaN         NaN          NaN   

      Cantidad  
4965     12667

In [5]:
homicidios_2018 = homicidios_2018.dropna()

In [6]:
#Updated dataset
homicidios_2018.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4965 entries, 0 to 4964
Data columns (total 20 columns):
Fecha                 4965 non-null object
Departamento          4965 non-null object
Municipio             4965 non-null object
Día                   4965 non-null object
Hora                  4965 non-null object
Barrio                4965 non-null object
Zona                  4965 non-null object
Clase de sitio        4965 non-null object
Arma empleada         4965 non-null object
Móvil Agresor         4965 non-null object
Móvil Victima         4965 non-null object
Edad                  4965 non-null float64
Sexo                  4965 non-null object
Estado civil          4965 non-null object
País de nacimiento    4965 non-null object
Clase de empleado     4965 non-null object
Profesión             4965 non-null object
Escolaridad           4965 non-null object
Código DANE           4965 non-null float64
Cantidad              4965 non-null int64
dtypes: float64(2), int64(1), ob

The first procedure is to have an overview of the data, I will do so by using the value_counts() method which gives us information of the data in the dataset for a desired column and the frequency of that value

In [7]:
#Checking if there are any duplicated values
homicidios_2018.duplicated().value_counts()

False    4965
dtype: int64

In [8]:
# dates which are consistently repeated and its frequency
homicidios_2018['Fecha'].value_counts()

01/01/2018 12:00:00 AM    82
08/07/2018 12:00:00 AM    72
08/04/2018 12:00:00 AM    67
08/12/2018 12:00:00 AM    67
12/08/2018 12:00:00 AM    65
11/02/2018 12:00:00 AM    64
03/06/2018 12:00:00 AM    62
02/09/2018 12:00:00 AM    60
09/09/2018 12:00:00 AM    58
05/08/2018 12:00:00 AM    56
04/02/2018 12:00:00 AM    55
02/12/2018 12:00:00 AM    54
09/06/2018 12:00:00 AM    54
01/07/2018 12:00:00 AM    52
11/11/2018 12:00:00 AM    51
06/05/2018 12:00:00 AM    50
09/12/2018 12:00:00 AM    48
04/11/2018 12:00:00 AM    47
04/03/2018 12:00:00 AM    47
07/10/2018 12:00:00 AM    47
11/06/2018 12:00:00 AM    46
10/11/2018 12:00:00 AM    45
08/09/2018 12:00:00 AM    44
01/09/2018 12:00:00 AM    44
10/02/2018 12:00:00 AM    43
02/06/2018 12:00:00 AM    43
06/10/2018 12:00:00 AM    42
03/07/2018 12:00:00 AM    41
07/07/2018 12:00:00 AM    40
09/07/2018 12:00:00 AM    40
                          ..
07/11/2018 12:00:00 AM    26
08/06/2018 12:00:00 AM    26
03/04/2018 12:00:00 AM    26
08/08/2018 12:

It can be seen that January 1st (Public holiday) is the day with the highest amount of homicides in Colombia in 2018, followed by July 8th (Sunday) and December 8th (Public holiday)
- 01/01/2018 12:00:00 AM    82
- 08/07/2018 12:00:00 AM    72
- 08/12/2018 12:00:00 AM    67

There is no data quality issues for this column, the data look consistent.

In [9]:
top3_dep = homicidios_2018['Departamento'].value_counts()
top3_dep

ANTIOQUIA             926
VALLE                 889
CUNDINAMARCA          581
CAUCA                 282
NARIÑO                263
ATLÁNTICO             215
NORTE DE SANTANDER    181
BOLÍVAR               156
CÓRDOBA               139
META                  116
CALDAS                100
HUILA                  99
CHOCÓ                  93
TOLIMA                 93
CESAR                  92
MAGDALENA              91
QUINDÍO                84
SANTANDER              83
RISARALDA              82
GUAJIRA                78
CAQUETÁ                70
PUTUMAYO               64
SUCRE                  58
ARAUCA                 52
BOYACÁ                 32
CASANARE               23
VICHADA                 6
GUAVIARE                5
SAN ANDRÉS              5
AMAZONAS                4
VAUPÉS                  2
GUAINÍA                 1
Name: Departamento, dtype: int64

In [10]:
total_top3_dep = top3_dep[0]+top3_dep[1]+top3_dep[2]
percentage_dep = total_top3_dep/4965
percentage_dep

0.48257804632426987

The departments with the highest homicide number are Antioquia (925), Valle (889) and Cundinamarca (581), respectively. These departments represent take 48% of the homicides in Colombia in 2018. There is no any data issues here.

In [11]:
homicidios_2018['Municipio'].value_counts()

CALI (CT)                  489
BOGOTÁ D.C. (CT)           431
MEDELLÍN (CT)              201
BARRANQUILLA (CT)          120
SAN ANDRES DE TUMACO        95
CARTAGENA (CT)              90
CÚCUTA (CT)                 65
SOACHA                      61
SOLEDAD                     57
PALMIRA                     57
CAUCASIA                    51
TURBO                       51
VILLAVICENCIO (CT)          49
QUIBDÓ (CT)                 42
TARAZÁ                      42
TULUÁ                       40
MANIZALES (CT)              39
NEIVA (CT)                  38
SANTA MARTA (CT)            38
YUMBO                       37
MAICAO                      36
MONTERÍA (CT)               36
PEREIRA (CT)                36
POPAYÁN (CT)                36
ARMENIA (CT)                34
JAMUNDÍ                     33
PASTO (CT)                  32
IBAGUÉ (CT)                 32
VALLEDUPAR (CT)             31
BUCARAMANGA (CT)            28
                          ... 
CHOCONTÁ                     1
INÍRIDA 

Cali, Bogota and Medellin are the cities with the highest number of homicides during 2018.

In [12]:
homicidios_2018['Día'].value_counts()

Domingo      1087
Sábado        812
Lunes         771
Jueves        610
Miércoles     581
Viernes       568
Martes        536
Name: Día, dtype: int64

A great number of homicides are committed on Sunday or Saturday. It is important to notice how monday is at the top 3 and friday is the second least day with homicides.

In [13]:
homicidios_2018['Hora'].value_counts().head()

20:30:00    80
20:00:00    79
21:30:00    71
23:00:00    65
19:30:00    64
Name: Hora, dtype: int64

In [14]:
homicidios_2018['Hora'].value_counts().tail()

21:37:00    1
20:24:00    1
6:12:00     1
8:32:00     1
8:11:00     1
Name: Hora, dtype: int64

The data looks consistent, although we can create subgroups to classify them and manage it in a different way.

In [15]:
homicidios_2018['Barrio'].value_counts()

CENTRO                                108
VEREDA                                 32
LA ESPERANZA                           19
EL BOSQUE                              19
EL CENTRO                              15
PARQUE PRINCIPAL                       15
OBRERO                                 13
EL VERGEL E13                          13
MOJICA E15                             12
EL PRADO                               12
SAN VICENTE                            12
LA CANDELARIA C-10                     11
POTRERO GRANDE E21                     11
REBOLO                                 11
CGTO. CURRULAO                         11
OLAYA HERRERA                          11
BARRIO NUEVO                           11
ALFONSO B. ARAGON E14                  10
EL TRIUNFO                             10
BUENOS AIRES                           10
EL PORVENIR                            10
LOS COMUNEROS I E15                    10
SAN FERNANDO                           10
SUCRE E9                          

'Centro' or the city center remains as the sector of the city where most homicides are committed

In [16]:
homicidios_2018['Zona'].value_counts()

URBANA    3321
RURAL     1644
Name: Zona, dtype: int64

This data seems pretty good to pass it through to the next stage.

In [17]:
homicidios_2018['Clase de sitio'].value_counts()

VIAS PUBLICAS                              2990
FINCAS Y SIMILARES                          561
CASAS DE HABITACION                         300
CARRETERA                                   144
FRENTE A RESIDENCIAS - VIA PUBLICA          131
BARES, CANTINAS Y SIMILARES                  78
LOTE BALDIO                                  70
RIOS                                         64
DENTRO DE LA VIVIENDA                        43
TIENDA                                       37
PARQUES                                      32
TROCHA                                       29
INTERIOR VEHICULO PARTICULAR                 26
BILLARES                                     25
SOBRE ANDEN - VIA PUBLICA                    20
TRAMO DE VIA                                 19
LOCAL COMERCIAL                              19
DISCOTECAS                                   17
CANCHA DE FUTBOL                             17
HOTELES, RESIDENCIAS, Y SIMILARES.           16
ESTABLECIMIENTO PUBLICO                 

In [18]:
homicidios_2018['Arma empleada'].value_counts()

ARMA DE FUEGO                         3549
ARMA BLANCA / CORTOPUNZANTE           1133
CONTUNDENTES                           217
CUERDA/SOGA/CADENA                      14
ARTEFACTO EXPLOSIVO/CARGA DINAMITA      14
CINTAS/CINTURON                          8
COMBUSTIBLE                              7
BOLSA PLASTICA                           6
MINA ANTIPERSONA                         5
NO REPORTADO                             3
ALMOHADA                                 3
GRANADA DE MANO                          3
MOTO BOMBA                               1
QUIMICOS                                 1
SIN EMPLEO DE ARMAS                      1
Name: Arma empleada, dtype: int64

In [19]:
homicidios_2018['Móvil Agresor'].value_counts()

A PIE                    4129
PASAJERO MOTOCICLETA      620
CONDUCTOR MOTOCICLETA     107
PASAJERO VEHICULO          29
PASAJERO TAXI              19
CONDUCTOR VEHICULO         18
BICICLETA                  16
-                          10
CONDUCTOR TAXI              7
PASAJERO BARCO              7
PASAJERO BUS                3
Name: Móvil Agresor, dtype: int64

In [20]:
homicidios_2018['Móvil Victima'].value_counts()

A PIE                    4553
CONDUCTOR MOTOCICLETA     185
CONDUCTOR VEHICULO         95
PASAJERO MOTOCICLETA       33
PASAJERO VEHICULO          27
CONDUCTOR TAXI             24
BICICLETA                  13
-                          12
PASAJERO TAXI               9
PASAJERO BARCO              8
PASAJERO BUS                4
CONDUCTOR BUS               2
Name: Móvil Victima, dtype: int64

The majority of homicides are commited by walk and the affected person also is nurdered while walking. There is no given data for these 2 as seen with the element with '-' in the tables above.

In [22]:
homicidios_2018['Edad'].describe()

count    4965.000000
mean       32.758711
std        13.108539
min         1.000000
25%        23.000000
50%        30.000000
75%        39.000000
max       100.000000
Name: Edad, dtype: float64

By analysing some statistics from Age, I found the minimum age was 1, which might be logical, however when displaying all details for this specific aspect, I found out that 'sort of Employee' for one of them was independent, then it clearly seems no logical.

In [23]:
homicidios_2018[homicidios_2018['Edad'] == 1]

Fecha Departamento       Municipio     Día     Hora  \
704   10/02/2018 12:00:00 AM      CAQUETÁ  FLORENCIA (CT)  Sábado  1:20:00   
2871  12/07/2018 12:00:00 AM    ANTIOQUIA        APARTADÓ  Jueves  0:00:00   

        Barrio    Zona                      Clase de sitio Arma empleada  \
704   RICAURTE  URBANA  HOTELES, RESIDENCIAS, Y SIMILARES.  CONTUNDENTES   
2871    OBRERO  URBANA                 CASAS DE HABITACION  CONTUNDENTES   

     Móvil Agresor Móvil Victima  Edad      Sexo Estado civil  \
704          A PIE         A PIE   1.0  FEMENINO      SOLTERO   
2871         A PIE         A PIE   1.0  FEMENINO      SOLTERO   

     País de nacimiento Clase de empleado Profesión   Escolaridad  \
704            COLOMBIA       DESEMPLEADO         -    ANALFABETA   
2871           COLOMBIA     INDEPENDIENTE         -  NO REPORTADO   

      Código DANE  Cantidad  
704    18001000.0         1  
2871    5045000.0         1

In [24]:
# Given that these details seem no ideal, then I condider deleting this row.
homicidios_2018 = homicidios_2018.drop(homicidios_2018.index[2871])
homicidios_2018.reset_index(drop=True)

Fecha        Departamento             Municipio  \
0     01/01/2018 12:00:00 AM            AMAZONAS          LETICIA (CT)   
1     01/01/2018 12:00:00 AM            AMAZONAS          LETICIA (CT)   
2     01/01/2018 12:00:00 AM           ANTIOQUIA                 ANDES   
3     01/01/2018 12:00:00 AM           ANTIOQUIA                 ANDES   
4     01/01/2018 12:00:00 AM           ANTIOQUIA                 ANORÍ   
5     01/01/2018 12:00:00 AM           ANTIOQUIA                 ANORÍ   
6     01/01/2018 12:00:00 AM           ANTIOQUIA                 ANORÍ   
7     01/01/2018 12:00:00 AM           ANTIOQUIA              CAUCASIA   
8     01/01/2018 12:00:00 AM           ANTIOQUIA              HISPANIA   
9     01/01/2018 12:00:00 AM           ANTIOQUIA                ITAGUI   
10    01/01/2018 12:00:00 AM           ANTIOQUIA         MEDELLÍN (CT)   
11    01/01/2018 12:00:00 AM           ANTIOQUIA         MEDELLÍN (CT)   
12    01/01/2018 12:00:00 AM           ANTIOQUIA                 TURBO   
13    01/01/2018 12:00:00 AM           ANTIOQUIA               VEGACHÍ   
14    01/01/2018 12:00:00 AM           ATLÁNTICO            CANDELARIA   
15    01/01/2018 12:00:00 AM           ATLÁNTICO               SOLEDAD   
16    01/01/2018 12:00:00 AM           ATLÁNTICO               SOLEDAD   
17    01/01/2018 12:00:00 AM             BOLÍVAR        CARTAGENA (CT)   
18    01/01/2018 12:00:00 AM             BOLÍVAR        CARTAGENA (CT)   
19    01/01/2018 12:00:00 AM             BOLÍVAR        CARTAGENA (CT)   
20    01/01/2018 12:00:00 AM             BOLÍVAR        CARTAGENA (CT)   
21    01/01/2018 12:00:00 AM             BOLÍVAR  EL CARMEN DE BOLÍVAR   
22    01/01/2018 12:00:00 AM              BOYACÁ              SOGAMOSO   
23    01/01/2018 12:00:00 AM              CALDAS        MANIZALES (CT)   
24    01/01/2018 12:00:00 AM            CASANARE  SAN LUIS DE PALENQUE   
25    01/01/2018 12:00:00 AM               CAUCA              EL TAMBO   
26    01/01/2018 12:00:00 AM               CAUCA                  PAEZ   
27    01/01/2018 12:00:00 AM               CAUCA          POPAYÁN (CT)   
28    01/01/2018 12:00:00 AM               CAUCA          POPAYÁN (CT)   
29    01/01/2018 12:00:00 AM               CAUCA          POPAYÁN (CT)   
...                      ...                 ...                   ...   
4934  12/12/2018 12:00:00 AM           ANTIOQUIA              CAUCASIA   
4935  12/12/2018 12:00:00 AM           ANTIOQUIA              CAUCASIA   
4936  12/12/2018 12:00:00 AM           ANTIOQUIA              ENVIGADO   
4937  12/12/2018 12:00:00 AM           ANTIOQUIA              ENVIGADO   
4938  12/12/2018 12:00:00 AM           ANTIOQUIA            LA PINTADA   
4939  12/12/2018 12:00:00 AM           ANTIOQUIA         MEDELLÍN (CT)   
4940  12/12/2018 12:00:00 AM             BOLÍVAR    SANTA ROSA DEL SUR   
4941  12/12/2018 12:00:00 AM               CAUCA              GUACHENÉ   
4942  12/12/2018 12:00:00 AM               CAUCA              GUACHENÉ   
4943  12/12/2018 12:00:00 AM               CAUCA                  PAEZ   
4944  12/12/2018 12:00:00 AM               CAUCA            VILLA RICA   
4945  12/12/2018 12:00:00 AM               CAUCA            VILLA RICA   
4946  12/12/2018 12:00:00 AM             CÓRDOBA            BUENAVISTA   
4947  12/12/2018 12:00:00 AM             CÓRDOBA            BUENAVISTA   
4948  12/12/2018 12:00:00 AM        CUNDINAMARCA      BOGOTÁ D.C. (CT)   
4949  12/12/2018 12:00:00 AM        CUNDINAMARCA      BOGOTÁ D.C. (CT)   
4950  12/12/2018 12:00:00 AM        CUNDINAMARCA              GIRARDOT   
4951  12/12/2018 12:00:00 AM        CUNDINAMARCA                SOACHA   
4952  12/12/2018 12:00:00 AM             GUAJIRA                MAICAO   
4953  12/12/2018 12:00:00 AM             GUAJIRA         RIOHACHA (CT)   
4954  12/12/2018 12:00:00 AM               HUILA                GARZÓN   
4955  12/12/2018 12:00:00 AM               HUILA              PITALITO   
4956  12/12/2018 12:00:00 A

In [25]:
(homicidios_2018['Edad'] >= 90).value_counts()

False    4963
True        1
Name: Edad, dtype: int64

In [26]:
homicidios_2018[homicidios_2018['Edad']>=80]

Fecha  Departamento            Municipio        Día  \
135   03/01/2018 12:00:00 AM     RISARALDA  SANTA ROSA DE CABAL  Miércoles   
648   08/02/2018 12:00:00 AM         VALLE            CALI (CT)     Jueves   
873   02/03/2018 12:00:00 AM        BOYACÁ             SOGAMOSO    Viernes   
987   05/03/2018 12:00:00 AM         CAUCA        PUERTO TEJADA      Lunes   
1115  09/03/2018 12:00:00 AM       CÓRDOBA       CIÉNAGA DE ORO    Viernes   
1600  11/04/2018 12:00:00 AM        ARAUCA            ARAUQUITA  Miércoles   
1717  03/05/2018 12:00:00 AM        CALDAS                NEIRA     Jueves   
1754  04/05/2018 12:00:00 AM         CESAR         PUEBLO BELLO    Viernes   
2336  09/06/2018 12:00:00 AM        TOLIMA         PURIFICACIÓN     Sábado   
2377  10/06/2018 12:00:00 AM         VALLE            CALI (CT)    Domingo   
2439  12/06/2018 12:00:00 AM  CUNDINAMARCA     BOGOTÁ D.C. (CT)     Martes   
2451  12/06/2018 12:00:00 AM         VALLE              PALMIRA     Martes   
2838  10/07/2018 12:00:00 AM         VALLE             RESTREPO     Martes   
3148  08/08/2018 12:00:00 AM         HUILA               TERUEL  Miércoles   
3198  10/08/2018 12:00:00 AM  CUNDINAMARCA     BOGOTÁ D.C. (CT)    Viernes   
3839  04/10/2018 12:00:00 AM         VALLE            CALI (CT)     Jueves   
4143  02/11/2018 12:00:00 AM         VALLE            CALI (CT)    Viernes   
4213  04/11/2018 12:00:00 AM        NARIÑO              IPIALES    Domingo   
4326  08/11/2018 12:00:00 AM         HUILA             PITALITO     Jueves   
4657  05/12/2018 12:00:00 AM     ANTIOQUIA             RIONEGRO  Miércoles   
4802  08/12/2018 12:00:00 AM     SANTANDER                GIRÓN     Sábado   

          Hora                    Barrio    Zona       Clase de sitio  \
135   20:50:00  CORREGIMIENTO LA CAPILLA   RURAL   FINCAS Y SIMILARES   
648   12:45:00    LOS CONQUISTADORES E11  URBANA         RESTAURANTES   
873   23:20:00                    CENTRO  URBANA  CASAS DE HABITACION   
987   10:30:00     CARLOS ALBERTO GUZMAN  URBANA        VIAS PUBLICAS   
1115  11:30:00                    CENTRO  URBANA               TIENDA   
1600   0:30:00              LAS PALMERAS  URBANA  CASAS DE HABITACION   
1717  14:03:00            VDA TROCADEROS   RURAL   FINCAS Y SIMILARES   
1754   7:20:00                  EL PRADO  URBANA  CASAS DE HABITACION   
2336   7:00:00    VRDA. VILLA ESPEREANZA   RURAL        VIAS PUBLICAS   
2377  11:30:00        POTRERO GRANDE E21  URBANA        VIAS PUBLICAS   
2439  20:01:00           SANTA ROSA E-10  URBANA  CASAS DE HABITACION   
2451  10:38:00               LA TRINIDAD  URBANA  CASAS DE HABITACION   
2838   9:20:00             VDA. CALIMITA   RURAL        VIAS PUBLICAS   
3148  16:30:00     VEREDA EL ALMORZADERO   RURAL  CASAS DE HABITACION   
3198   9:18:00         LAS ORQUÍDEAS E-1  URBANA        VIAS PUBLICAS   
3839  23:00:00            LA SULTANA E20  URBANA        VIAS PUBLICAS   
4143  14:40:00            SAN NICOLAS E3  URBANA        VIAS PUBLICAS   
4213  18:00:00                 CHAMPAGTH  URBANA        VIAS PUBLICAS   
4326  10:00:00                 GUADUALES  URBANA  CASAS DE HABITACION   
4657  17:30:00                EL LLANITO  URBANA        VIAS PUBLICAS   
4802  18:00:00        HACIENDA LA MESETA  URBANA  CASAS DE HABITACION   

                    Arma empleada Móvil Agresor Móvil Victima   Edad  \
135                  CONTUNDENTES         A PIE         A PIE   83.0   
648                 ARMA DE FUEGO         A PIE         A PIE   83.0   
873                  CONTUNDENTES         A PIE         A PIE   80.0   
987                 ARMA DE FUEGO         A PIE         A PIE   83.0   
1115  ARMA BLANCA / CORTOPUNZANTE         A PIE         A PIE   81.0   
1600  ARMA BLANCA / CORTOPUNZANTE         A PIE         A PIE   82.0   
1717           CUERDA/SOGA/CADENA         A PIE         A PIE   88.0   
1754  ARMA BLANCA / CORTOPUNZANTE         A PIE         A PIE   86.0   
2336                ARMA DE FUEGO         A PIE    

In [27]:
homicidios_2018['Sexo'].value_counts()

MASCULINO    4555
FEMENINO      409
Name: Sexo, dtype: int64

In [28]:
homicidios_2018['Estado civil'].value_counts()

SOLTERO        3141
UNION LIBRE    1502
CASADO          241
SEPARADO         38
DIVORCIADO       17
-                13
VIUDO            12
Name: Estado civil, dtype: int64

In [29]:
homicidios_2018[homicidios_2018['Estado civil'] == '-']

Fecha        Departamento          Municipio  \
870   02/03/2018 12:00:00 AM           ANTIOQUIA              TURBO   
982   05/03/2018 12:00:00 AM           ANTIOQUIA              YONDÓ   
2213  06/06/2018 12:00:00 AM           ANTIOQUIA        LA ESTRELLA   
2630  05/07/2018 12:00:00 AM  NORTE DE SANTANDER        CÚCUTA (CT)   
2631  05/07/2018 12:00:00 AM  NORTE DE SANTANDER        CÚCUTA (CT)   
2941  02/08/2018 12:00:00 AM              NARIÑO          SAMANIEGO   
3098  06/08/2018 12:00:00 AM             GUAJIRA             MAICAO   
3164  09/08/2018 12:00:00 AM           ANTIOQUIA      MEDELLÍN (CT)   
3176  09/08/2018 12:00:00 AM  NORTE DE SANTANDER  VILLA DEL ROSARIO   
3447  03/09/2018 12:00:00 AM           SANTANDER   BUCARAMANGA (CT)   
3819  04/10/2018 12:00:00 AM              ARAUCA               TAME   
3852  05/10/2018 12:00:00 AM           ANTIOQUIA      MEDELLÍN (CT)   
4785  08/12/2018 12:00:00 AM        CUNDINAMARCA             SOACHA   

            Día      Hora                              Barrio    Zona  \
870     Viernes  19:40:00                VDA. BOCAS DE ATRATO   RURAL   
982       Lunes  11:30:00                    VDA. SANTA CLARA   RURAL   
2213  Miércoles  19:00:00                     VDA. EL ROMERAL   RURAL   
2630     Jueves   8:30:00              BARRIO LLERAS RESTREPO  URBANA   
2631     Jueves   4:45:00                 BARRIO CUBEROS NIÑO  URBANA   
2941     Jueves  20:00:00                   VEREDA EL MOTILON   RURAL   
3098      Lunes   6:45:00     TRAMO DE VIA CARRAIPIA - MAICAO  URBANA   
3164     Jueves  13:53:00                   BARRIO COLON C-10  URBANA   
3176     Jueves   9:30:00  SECTOR LA PLAYA(VILLA DEL ROSARIO)  URBANA   
3447      Lunes   1:41:00                             ALARCON  URBANA   
3819     Jueves  16:30:00                   VEREDA EL BOTALON   RURAL   
3852    Viernes  23:41:00                  LA CANDELARIA C-10  URBANA   
4785     Sábado  11:00:00                     LAGOS DE MALIBU  URBANA   

                 Clase de sitio                Arma empleada  \
870          FINCAS Y SIMILARES                ARMA DE FUEGO   
982               VIAS PUBLICAS                ARMA DE FUEGO   
2213              VIAS PUBLICAS                ARMA DE FUEGO   
2630               CANALIZACION  ARMA BLANCA / CORTOPUNZANTE   
2631              VIAS PUBLICAS                ARMA DE FUEGO   
2941              VIAS PUBLICAS                ARMA DE FUEGO   
3098                LOTE BALDIO                  COMBUSTIBLE   
3164  SOBRE ANDEN - VIA PUBLICA                ARMA DE FUEGO   
3176                     TROCHA                ARMA DE FUEGO   
3447              VIAS PUBLICAS  ARMA BLANCA / CORTOPUNZANTE   
3819                 CEMENTERIO                ARMA DE FUEGO   
3852              VIAS PUBLICAS  ARMA BLANCA / CORTOPUNZANTE   
4785              VIAS PUBLICAS  ARMA BLANCA / CORTOPUNZANTE   

             Móvil Agresor          Móvil Victima  Edad       Sexo  \
870                  A PIE                  A PIE  22.0  MASCULINO   
982                  A PIE                  A PIE  55.0  MASCULINO   
2213                 A PIE                  A PIE  30.0  MASCULINO   
2630                 A PIE                  A PIE  28.0  MASCULINO   
2631                 A PIE                  A PIE  37.0  MASCULINO   
2941                 A PIE  CONDUCTOR MOTOCICLETA  29.0  MASCULINO   
3098                 A PIE                  A PIE  38.0  MASCULINO   
3164                 A PIE                  A PIE  30.0  MASCULINO   
3176                 A PIE                  A PIE  25.0  MASCULINO   
3447                 A PIE                  A PIE  35.0  MASCULINO   
3819  PASAJERO MOTOCICLETA                  A PIE  26.0  MASCULINO   
3852                 A PIE                  A PIE  35.0  MASCULINO   
4785                 A PIE                  A PIE  25.0  MASCULINO   

     Estado civil País de nacimiento Clase de empleado Profesión Escolaridad  \
870             -                  -        NO 

By analysing 'Estado civil' for missing data, I consider there is no need for deleting these missing rows, these do not represent a critical feature at the momement.

In [30]:
homicidios_2018['País de nacimiento'].value_counts()

COLOMBIA     4828
VENEZUELA      91
-              35
ECUADOR         3
BRASIL          2
PERU            1
ALEMANIA        1
ESPAÑA          1
NORUEGA         1
ITALIA          1
Name: País de nacimiento, dtype: int64

In [31]:
homicidios_2018['Clase de empleado'].value_counts()

EMPLEADO PARTICULAR       1654
INDEPENDIENTE             1515
DESEMPLEADO                847
AGRICULTOR                 331
COMERCIANTE                156
ESTUDIANTE                 131
AMA DE CASA                111
EMPLEADO POLICIAL           36
PENSIONADO                  32
DELINCUENCIA                30
EMPLEADO EJERCITO           28
ETNIA INDIGENA              15
LIDER CIVICO                14
NO REPORTA                  13
EMPLEADO PUBLICO            12
GRUPOS ILEGALES             12
EDUCADOR                     7
GANADERO                     5
EMPLEADO INPEC               4
EMPLEADO CTI                 3
EMPLEADO ARMADA/MARINA       2
DEPORTISTA                   2
EMPLEADO SALUD               2
AFRODESCENDIENTE             1
POLITICO                     1
Name: Clase de empleado, dtype: int64

'Profesion' is a critical variable given that more than 98% of data contains this piece of information and there is also no reported data whitin this group, this variable does not allow us to develop a proper data analysis

In [32]:
homicidios_2018['Profesión'].value_counts()

-                                                          4837
NO REPORTADA                                                 47
TECNICO PROFESIONAL EN SERVICIO DE POLICIA                   24
POLICIAS                                                     13
DERECHO                                                       4
LICENCIADO                                                    4
INGENIERO                                                     4
COMUNICACION SOCIAL- PERIODISMO                               2
INGENIERÍA INDUSTRIAL                                         2
CONTADURIA PUBLICA                                            2
TECNICO PROFESIONAL EN ADMINISTRACION  DE EMPRESAS            2
ADMINISTRACION  DE EMPRESAS                                   2
ADMINISTRACION  COMERCIAL                                     2
INGENIERÍA MECANICA                                           1
TECNICO PROFESIONAL EN DOCENCIA INDUSTRIAL ELECTRICIDAD       1
PROFESORES DE EDUCACION MEDIA           

In [33]:
(homicidios_2018['Profesión'] == '-').value_counts()

True     4837
False     127
Name: Profesión, dtype: int64

In [34]:
homicidios_2018['Escolaridad'].value_counts()

SECUNDARIA      2936
PRIMARIA        1709
ANALFABETA       107
TECNICO           73
SUPERIOR          68
NO REPORTADO      53
-                 14
TECNOLOGO          4
Name: Escolaridad, dtype: int64

In [35]:
homicidios_2018['Código DANE'].value_counts()

76001000.0    489
11001000.0    431
5001000.0     201
8001000.0     120
52835000.0     95
13001000.0     90
54001000.0     65
25754000.0     61
8758000.0      57
76520000.0     57
5837000.0      51
5154000.0      51
50001000.0     49
5790000.0      42
27001000.0     42
76834000.0     40
17001000.0     39
41001000.0     38
47001000.0     38
76892000.0     37
19001000.0     36
66001000.0     36
44430000.0     36
23001000.0     36
63001000.0     34
76364000.0     33
52001000.0     32
73001000.0     32
20001000.0     31
19573000.0     28
             ... 
63690000.0      1
13490000.0      1
13433000.0      1
66318000.0      1
25035000.0      1
70670000.0      1
17433000.0      1
25407000.0      1
52696000.0      1
27135000.0      1
50287000.0      1
50606000.0      1
25286000.0      1
81220000.0      1
70717000.0      1
25019000.0      1
19513000.0      1
73854000.0      1
13873000.0      1
23182000.0      1
52390000.0      1
5467000.0       1
20295000.0      1
20032000.0      1
25799000.0

In [36]:
(homicidios_2018['Código DANE'] == '-').value_counts()

False    4964
Name: Código DANE, dtype: int64

'Cantidad' is a variable which is not quite clear given that the data reported has variables which describes only 1 person, so I would rather not considering this feature for our analysis

In [37]:
homicidios_2018['Cantidad'].value_counts()

1    4956
2       7
3       1
Name: Cantidad, dtype: int64

In [38]:
homicidios_2018[homicidios_2018['Cantidad'] == 3]

Fecha Departamento Municipio     Día      Hora  \
1569  10/04/2018 12:00:00 AM        CAUCA     PATÍA  Martes  18:00:00   

           Barrio   Zona Clase de sitio  Arma empleada Móvil Agresor  \
1569  VDA BETANIA  RURAL    LOTE BALDIO  ARMA DE FUEGO         A PIE   

     Móvil Victima  Edad       Sexo Estado civil País de nacimiento  \
1569         A PIE  35.0  MASCULINO      SOLTERO           COLOMBIA   

     Clase de empleado Profesión Escolaridad  Código DANE  Cantidad  
1569       DESEMPLEADO         -    PRIMARIA   19532000.0         3

In [39]:
homicidios_2018.drop('Cantidad', axis=1, inplace=True)

In [40]:
homicidios_2018.drop('Profesión', axis=1, inplace=True)

# Final dataset for data analysis

To have a better data handling

In [41]:
homicidios_2018.rename( columns = { 'Fecha' : 'FECHA', 'Departamento' : 'DEPARTAMENTO', 'Municipio' : 'MUNICIPIO',
                                   'Día' : 'DIA', 'Hora' : 'HORA', 'Clase de sitio' : 'CLASE SITIO', 'Barrio' : 'BARRIO',
                                   'Zona' : 'ZONA', 'Arma empleada' : 'ARMA EMPLEADA', 'Móvil Agresor' : 'MOVIL AGRESOR', 
                                   'Móvil Victima' : 'MOVIL VICTIMA', 'Edad' : 'EDAD', 'Sexo' : 'GENERO',
                                   'Estado civil' : 'ESTADO CIVIL', 'País de nacimiento' : 'PAIS NACIMIENTO',
                                   'Clase de empleado' : 'CLASE EMPLEADO', 'Escolaridad' : 'ESCOLARIDAD',
                                   'Código DANE' : 'CODIGO DANE'}, inplace = True )
homicidios_2018['DIA'].replace({ 'Lunes' : 'LUNES', 'Martes' : 'MARTES', 'Miércoles' : 'MIERCOLES', 'Jueves' : 'JUEVES',
                              'Viernes' : 'VIERNES', 'Sábado' : 'SABADO', 'Domingo' : 'DOMINGO'}, inplace = True)
homicidios_2018.replace({'-' : 'NO REPORTADO'}, inplace = True)

In [42]:
homicidios_2018

FECHA        DEPARTAMENTO             MUNICIPIO  \
0     01/01/2018 12:00:00 AM            AMAZONAS          LETICIA (CT)   
1     01/01/2018 12:00:00 AM            AMAZONAS          LETICIA (CT)   
2     01/01/2018 12:00:00 AM           ANTIOQUIA                 ANDES   
3     01/01/2018 12:00:00 AM           ANTIOQUIA                 ANDES   
4     01/01/2018 12:00:00 AM           ANTIOQUIA                 ANORÍ   
5     01/01/2018 12:00:00 AM           ANTIOQUIA                 ANORÍ   
6     01/01/2018 12:00:00 AM           ANTIOQUIA                 ANORÍ   
7     01/01/2018 12:00:00 AM           ANTIOQUIA              CAUCASIA   
8     01/01/2018 12:00:00 AM           ANTIOQUIA              HISPANIA   
9     01/01/2018 12:00:00 AM           ANTIOQUIA                ITAGUI   
10    01/01/2018 12:00:00 AM           ANTIOQUIA         MEDELLÍN (CT)   
11    01/01/2018 12:00:00 AM           ANTIOQUIA         MEDELLÍN (CT)   
12    01/01/2018 12:00:00 AM           ANTIOQUIA                 TURBO   
13    01/01/2018 12:00:00 AM           ANTIOQUIA               VEGACHÍ   
14    01/01/2018 12:00:00 AM           ATLÁNTICO            CANDELARIA   
15    01/01/2018 12:00:00 AM           ATLÁNTICO               SOLEDAD   
16    01/01/2018 12:00:00 AM           ATLÁNTICO               SOLEDAD   
17    01/01/2018 12:00:00 AM             BOLÍVAR        CARTAGENA (CT)   
18    01/01/2018 12:00:00 AM             BOLÍVAR        CARTAGENA (CT)   
19    01/01/2018 12:00:00 AM             BOLÍVAR        CARTAGENA (CT)   
20    01/01/2018 12:00:00 AM             BOLÍVAR        CARTAGENA (CT)   
21    01/01/2018 12:00:00 AM             BOLÍVAR  EL CARMEN DE BOLÍVAR   
22    01/01/2018 12:00:00 AM              BOYACÁ              SOGAMOSO   
23    01/01/2018 12:00:00 AM              CALDAS        MANIZALES (CT)   
24    01/01/2018 12:00:00 AM            CASANARE  SAN LUIS DE PALENQUE   
25    01/01/2018 12:00:00 AM               CAUCA              EL TAMBO   
26    01/01/2018 12:00:00 AM               CAUCA                  PAEZ   
27    01/01/2018 12:00:00 AM               CAUCA          POPAYÁN (CT)   
28    01/01/2018 12:00:00 AM               CAUCA          POPAYÁN (CT)   
29    01/01/2018 12:00:00 AM               CAUCA          POPAYÁN (CT)   
...                      ...                 ...                   ...   
4935  12/12/2018 12:00:00 AM           ANTIOQUIA              CAUCASIA   
4936  12/12/2018 12:00:00 AM           ANTIOQUIA              CAUCASIA   
4937  12/12/2018 12:00:00 AM           ANTIOQUIA              ENVIGADO   
4938  12/12/2018 12:00:00 AM           ANTIOQUIA              ENVIGADO   
4939  12/12/2018 12:00:00 AM           ANTIOQUIA            LA PINTADA   
4940  12/12/2018 12:00:00 AM           ANTIOQUIA         MEDELLÍN (CT)   
4941  12/12/2018 12:00:00 AM             BOLÍVAR    SANTA ROSA DEL SUR   
4942  12/12/2018 12:00:00 AM               CAUCA              GUACHENÉ   
4943  12/12/2018 12:00:00 AM               CAUCA              GUACHENÉ   
4944  12/12/2018 12:00:00 AM               CAUCA                  PAEZ   
4945  12/12/2018 12:00:00 AM               CAUCA            VILLA RICA   
4946  12/12/2018 12:00:00 AM               CAUCA            VILLA RICA   
4947  12/12/2018 12:00:00 AM             CÓRDOBA            BUENAVISTA   
4948  12/12/2018 12:00:00 AM             CÓRDOBA            BUENAVISTA   
4949  12/12/2018 12:00:00 AM        CUNDINAMARCA      BOGOTÁ D.C. (CT)   
4950  12/12/2018 12:00:00 AM        CUNDINAMARCA      BOGOTÁ D.C. (CT)   
4951  12/12/2018 12:00:00 AM        CUNDINAMARCA              GIRARDOT   
4952  12/12/2018 12:00:00 AM        CUNDINAMARCA                SOACHA   
4953  12/12/2018 12:00:00 AM             GUAJIRA                MAICAO   
4954  12/12/2018 12:00:00 AM             GUAJIRA         RIOHACHA (CT)   
4955  12/12/2018 12:00:00 AM               HUILA                GARZÓN   
4956  12/12/2018 12:00:00 AM               HUILA              PITALITO   
4957  12/12/2018 12:00:00 A

## In conlcusion for point 2, there has been a profound analysis of each of the elements of the dataset, this is very sensible data considering that it involves homicides data of Colombia, therefore I have found two very critical variables which were deleted from the final table, these are 'Profesion' and 'Cantidad'. I consider, it is important to have clear data and the source of it should be contacted if there are misleading information. 

## There is another aspect which is important to consider and it is the '-' or not given data, although I do not see it as a critical aspect to leave it out of the dataset, it should be considered if there is further analysis where those become an important part of it.

In [43]:
homicidios_2018.to_csv('./modified_dataset/homicidios_2018.csv', index=False)